### PR TITLE
Allow arbitrary dll extensions

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This is to help us collect information about the relevance of filename extensions when loading DLLs. Before this PR, things only work if assembly names end with `.dll` because the runtime requires that (it resolves assembly references using the filenames we supply). After this PR, you can load assemblies from URLs containing any filename extension.

This is not to imply that the URLs *should* change from `.dll` to anything else; it's merely to make it possible for people who want to patch their publish output manually for any reason. This means we have a way to test whether extension-changing is beneficial in any real-world scenario.

I suggest this goes into preview 2 because it's a super low risk change, and that helps us collect info sooner.

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
With this change, it's possible to modify the `.dll` extensions in the published output using the following instructions, which work for both "standalone" and "hosted" variants of Blazor WebAssembly projects:

### Windows

* Open a PowerShell prompt in the `wwwroot` directory of your published output (for example, in `yourapp\bin\Release\netstandard2.1\publish\wwwroot`
* Execute the following two lines:

```
dir .\_framework\_bin | rename-item -NewName { $_.name -replace ".dll\b",".bin" }
((Get-Content .\_framework\blazor.boot.json -Raw) -replace '.dll"','.bin"') | Set-Content .\_framework\blazor.boot.json
```

### Linux / macOS

* Open a terminal in the `wwwroot` directory of your published output (for example, in `yourapp/bin/Release/netstandard2.1/publish/wwwroot`
* Execute the following two lines:

```
for f in _framework/_bin/*; do mv "$f" "`echo $f | sed -e 's/\.dll\b/.bin/g'`"; done
sed -i 's/\.dll"/.bin"/g' _framework/blazor.boot.json
```

**Review by javiercn — APPROVED** [review]
I don't like it, but I think there's no better way to achieve this, so I'm ok with it.

As an alternative we could supply an option to pass in the pre-downloaded dlls that we wanted to use for the app in `Blazor.start` and that would give people more flexibility to not just change the extension but to also transform the contents in any way they saw fit for their needs without us having to know about it.

Does that make sense?

**Comment by SteveSandersonMS** [comment]
> I don't like it, but I think there's no better way to achieve this, so I'm ok with it.

That's basically my feeling too. If people find this practically fixes real-world problems then it gives us a reason to take further actions. If it doesn't, then no harm is done since this doesn't meaningfully change the API surface or feature set.

> As an alternative we could supply an option to pass in the pre-downloaded dlls

Yes, we could do that, but let's keep the change minimal until the reasons become clearer. We can expand the functionality to do what you're suggesting in the future if we need.

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 1]
```diff
@@ -198,11 +198,13 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
     MONO.loaded_files = [];
 
     // Fetch the assemblies and PDBs in the background, telling Mono to wait until they are loaded
+    // Mono requires the assembly filenames to have a '.dll' extension, so supply such names regardless
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts]: .NET WebAssembly

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts]: Sorry, already merged! I doubt this change warrants a further PR so will leave it, but if you disagree let me know and I'll do it.

**Pull request merged**